File: tests/e2e/reduce.py
# RUN: uv run xdsl-exo -o %t.mlir %s
# RUN: mlir-opt --convert-arith-to-llvm --convert-cf-to-llvm --convert-func-to-llvm --reconcile-unrealized-casts %t.mlir | mlir-translate --mlir-to-llvmir | llc -filetype=obj -o %t.o
# RUN: clang %t.o %S/reduce.c -o %t.bin -lm
# RUN: %t.bin | filecheck %s

from __future__ import annotations

from exo import *


# CHECK: 36.0
# CHECK: 36
@proc
def reduce_float(x: f32[8] @ DRAM, y: f32[1] @ DRAM):
    for i in seq(0, 8):
        y[0] += x[i]


@proc
def reduce_int(x: i32[8] @ DRAM, y: i32[1] @ DRAM):
    for i in seq(0, 8):
        y[0] += x[i]

File: tests/e2e/reduce.c
#include <stdio.h>

extern void reduce_float(float *x, float *y);
extern void reduce_int(int *x, int *y);

int main() {
    float xf[8] = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f, 8.0f};
    float yf[1] = {0.0f};
    reduce_float(xf, yf);
    printf("%.1f\n", yf[0]);

    int xi[8] = {1, 2, 3, 4, 5, 6, 7, 8};
    int yi[1] = {0};
    reduce_int(xi, yi);
    printf("%d\n", yi[0]);

    return 0;
}
